### PR TITLE
batch - support integration-name instead of integration-ID

### PR DESCRIPTION
**Pull request by OrNovo**
e.g. -

`integrationName: WebhookAlerts` instead of `integrationID: 2235`

### CODE REVIEW

File: controllers/alphacontrollers/alert_controller.go
Comment (OrNovo) [inline controllers/alphacontrollers/alert_controller.go:1000]: not sure how we want to handle APIs errors

File: controllers/alphacontrollers/alert_controller.go
Comment (OrNovo) [inline controllers/alphacontrollers/alert_controller.go:1004]: Also not sure about this error handling

File: apis/coralogix/v1alpha1/alert_types.go
Comment (nicolastakashi) [inline apis/coralogix/v1alpha1/alert_types.go:1064]: Not required, but I'd suggest getIntegration, findIntegration or serachIntegration

File: controllers/alphacontrollers/alert_controller.go
Comment (nicolastakashi) [inline controllers/alphacontrollers/alert_controller.go:999]: Should we avoid context.BackGround()?
And get the context from the caller?

File: controllers/alphacontrollers/alert_controller.go
Comment (nicolastakashi) [inline controllers/alphacontrollers/alert_controller.go:997]: Should this still exist, even if we're removing the integration id?

Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:997]: It suppose to be nil when you remove it

**Comment by OrNovo** [comment]
@nicolastakashi I guess charts also have to be updated?

**Comment by nicolastakashi** [comment]
> @nicolastakashi I guess charts also have to be updated?

We need to update the CRDs and also the Image version.
Worth we have a task to automate this soon

**Comment by OrNovo** [comment]
> > @nicolastakashi I guess charts also have to be updated?
> 
> We need to update the CRDs and also the Image version. Worth we have a task to automate this soon

so to leave it to different PR?

File: apis/coralogix/v1alpha1/alert_types.go
Comment (nicolastakashi) [inline apis/coralogix/v1alpha1/alert_types.go:1064]: Just wondering if maybe makes sense for this function to return a integration struct, so in the future, we can leverage this method to manage integrations using the Operator.

Comment (OrNovo) [reply apis/coralogix/v1alpha1/alert_types.go:1064]: When we mange integrations we (probably) won't have to query all the webhooks

File: controllers/alphacontrollers/alert_controller.go
Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:1000]: If we don't have think kind of knowledge, I guess it is fine to fail the reconciliation

File: controllers/alphacontrollers/alert_controller.go
Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:1004]: I think flattenNotification should return errors, and the error will be handled in the reconciliation loop

**Review by ArthurSens — COMMENTED** [review]
Awesome Or, thanks for tackling this one!

I left a few comments regarding error handling. It should be good to merge after that

**Comment by ArthurSens** [comment]
Also food for thought for a future improvement, we could avoid multiple calls to the webhook API if we do a single one when the operator is initializing and cache those webhooks in memory.

The operator itself can provide the webhook IDs, instead of calling the Webhook API once per Alert resource

File: controllers/alphacontrollers/alert_controller.go
Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:1000]: I think it would be better to leave it to reconciliation anyway to avoid complicate logic here

Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:1000]: Yeah, let's just return the error and decide what to do with it during reconciliation

File: controllers/alphacontrollers/alert_controller.go
Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:1004]: The purpose of flatten is to initialize or update the status of the alert. I don't want to fail the update of the status just because of a failure in placing the integration-name. How about leaving the integration-name nil?

Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:1004]: If we leave the integration name as nil, the operator will try to reconcile it again since the Spec and Status are different, right? The loop will continue forever, until we don't fail to fetch a valid `alias`.

I think we want to fail the status update here, it will continue this reconciliation loop, but at least the metric `reconciliation_errors_total` provided by kubebuilder will increase, and we can trigger alerts for this scenario

Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:1004]: The reconcile happens forever anyway, the question is just in witch period.
If the Spec and Status are different it will try to update the alert with Update call to the backend but the period to the next reconciliation stays the same.

Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:1004]: Thanks for the explanation :) 

Still, I believe we want to fail here so we can monitor this failures with kubebuilder built-in metrics

Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:1026]: It makes the code little bit ugly :'(  but let me know if you think this is a good solution

**Comment by OrNovo** [comment]
> Also food for thought for a future improvement, we could avoid multiple calls to the webhook API if we do a single one when the operator is initializing and cache those webhooks in memory.
> 
> The operator itself can provide the webhook IDs, instead of calling the Webhook API once per Alert resource

TBH I'm not a big fan of handling states inside the operator, but I guess we won't have a choice.

File: controllers/alphacontrollers/alert_controller.go
Comment (ArthurSens) [inline controllers/alphacontrollers/alert_controller.go:220]:
```suggestion
		log.Error(flattenErr, "Received an error while flattened Alert")
```

File: controllers/alphacontrollers/alert_controller.go
Comment (ArthurSens) [inline controllers/alphacontrollers/alert_controller.go:984]: nit:

Just in case you think this would be better, there are a few libraries that deal with concatenating multiple errors, for example:
* https://pkg.go.dev/github.com/hashicorp/go-multierror
* https://pkg.go.dev/errors

We could get rid of this if/for in case it sounds useful to you

**Review by ArthurSens — APPROVED** [review]
Awesome Or! Thanks for tackling this so quickly

I've added just a few more comments, but they are non-blocking. Feel free to merge or to address them, it is up to you :)

File: controllers/alphacontrollers/alert_controller.go
Comment (ArthurSens) [reply controllers/alphacontrollers/alert_controller.go:220]: We could also move this upwards, and return the error even before we try to call Status.Update

Comment (OrNovo) [reply controllers/alphacontrollers/alert_controller.go:220]: I prefer the status to be updated with nil integration-name

**Pull request merged**
